utils/shared/src/main/scala/fastparse/ParserInput.scala

```diff
@@ -0,0 +1,92 @@
+package fastparse
+
+import scala.collection.mutable.ArrayBuffer
+
+abstract class ParserInputLogger {
+  def logDropBuffer(index: Int)
+  def logApply(index: Int)
+}
+
+abstract class ParserInput[ElemType] {
+  def apply(index: Int): ElemType
+  def dropBuffer(index: Int): Unit
+  def slice(from: Int, until: Int): IndexedSeq[ElemType]
+  def length: Int
+  def innerLength: Int
+  def isReachable(index: Int): Boolean
+}
```

Comment:
This should be stable enough to document all these things. Apart from what everything does, some subtleties like `why do we need dropBuffer?` or `what's the difference between length and innerLength?` and `what is isReachable and what do we use it for?` is worth spending extra effort documenting